my-sms/spec/controllers/students_controller_spec.rb

```diff
@@ -23,119 +25,115 @@
+
       it "returns a success response (i.e. to display the 'new' template)" do
-        post :create, {:student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
+        should render_template('new')
       end
     end
   end
 
-  describe "PUT #update" do
-    context "with valid params" do
-      let(:new_attributes) {
-        skip("Add a hash of attributes valid for your model")
-      }
+  describe 'PUT #update' do
+    let(:student) { FactoryBot.create(:student) }
+    subject { put :update, { id: student.to_param, student: new_attributes } }
+
+    context 'with valid params' do
+      let(:new_attributes) { FactoryBot.attributes_for(:student) }
 
-      it "updates the requested student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => new_attributes}, valid_session
+      it 'updates the requested student' do
+        subject
         student.reload
-        skip("Add assertions for updated state")
+        attrs = %i[first_name last_name gender]
+        student_attrs = student.attributes.fetch_values(*attrs.map(&:to_s))
+        new_attrs = new_attributes.fetch_values(*attrs)
+        expect(student_attrs).to eq new_attrs
       end
 
-      it "redirects to the student" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => valid_attributes}, valid_session
-        expect(response).to redirect_to(student)
+      it 'redirects to the student' do
+        should be_redirect
+        should redirect_to(student)
       end
     end
 
-    context "with invalid params" do
+    context 'with invalid params' do
+      let(:new_student_args) { { first_name: nil } }
+      let(:new_attributes) do
+        FactoryBot.attributes_for(:student, { first_name: nil })
+      end
+
       it "returns a success response (i.e. to display the 'edit' template)" do
-        student = Student.create! valid_attributes
-        put :update, {:id => student.to_param, :student => invalid_attributes}, valid_session
-        expect(response).to be_success
+        should be_success
+        should render_template('edit')
       end
     end
   end
 
-  describe "DELETE #destroy" do
-    it "destroys the requested student" do
-      student = Student.create! valid_attributes
-      expect {
-        delete :destroy, {:id => student.to_param}, valid_session
-      }.to change(Student, :count).by(-1)
+  describe 'DELETE #destroy' do
+    let(:student) { FactoryBot.create(:student) }
+    subject do
+      delete :destroy, { id: student.to_param }
     end
 
-    it "redirects to the students list" do
-      student = Student.create! valid_attributes
-      delete :destroy, {:id => student.to_param}, valid_session
-      expect(response).to redirect_to(students_url)
+    it 'destroys the requested student' do
+      expect { subject }.to change(Student, :count).by(0)
```

Comment:
This _is_ testing that the delete works, as we know that a student record will be created when referenced. However, it's a misleading spec to read, as it appears that it's testing that the call has no effect.

What do you think of this instead?

```ruby
# Reference the student, so we know they're in the DB before the delete call
# You can do this in a before block, or in the first line of your `it` block.
before { student }

it 'destroys the requested student' do
  expect { subject }.to change(Student, :count).by(-1)
end
```

Reply:
This is much clearer. Added in b7da5ea777f1c6443f26dfaacc81841bf07d8888